### PR TITLE
Set `--configPath` flag for deployments

### PR DESCRIPTION
**Pull request by hemanik**
Adds `--configPath` flag  and set the default to `""`  for the following deployments: 

- Deck
- Hook
- Horologium
- Plank
- Sinker
- Tide

### CODE REVIEW

**Comment by cjwagner** [comment]
/cc @stevekuznetsov 
What do you think of this?

**Comment by fejta** [comment]
I also think it might make sense to first set these values in our cluster files, add an annoucement as well as a warning message in various logs.

And only a month or so after that actually set the default value to ""

**Comment by stevekuznetsov** [comment]
Y'all are too quick. Totally agree -- love the change, but would like to see:

1. change flags to be explicit
2. add a warning to the code if it's being set implicitly
3. TODO to nix the default in the future

File: prow/ANNOUNCEMENTS.md [hunk 1]
```diff
@@ -55,6 +55,8 @@ Note: versions specified in these announcements may not include bug fixes made
 in more recent versions so it is recommended that the most recent versions are
 used when updating deployments.
 
+ - *March 29, 2019* `hook`, `deck`, `horologium`, `tide`, `plank` and `sinker` will no longer provide a default value for the `--config-path` flag.
```
Comment (stevekuznetsov) [inline prow/ANNOUNCEMENTS.md]: nit: can you follow the line break convention here

File: prow/cmd/hook/main.go [hunk 2]
```diff
@@ -96,6 +96,11 @@ func main() {
 	}
 	logrus.SetFormatter(logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "hook"}))
 
+	if o.configPath == "" {
+		logrus.Warning("Hook requires explicitly setting --configPath to function correctly in production.")
+		logrus.Warning("--configPath will soon default to empty string.")
```
Comment (stevekuznetsov) [inline prow/cmd/hook/main.go]: Could you add a time to this please

Comment (hemanik) [reply prow/cmd/hook/main.go]: Could you suggest a time for this? Also, should I undo the changes for the default and add TODOs for the time being?

Comment (stevekuznetsov) [reply prow/cmd/hook/main.go]: yes, we should not change the defaults in this PR. 6months is a good time frame to sunset this default

File: prow/cmd/plank/main.go [hunk 3]
```diff
@@ -101,6 +101,11 @@ func main() {
 
 	pjutil.ServePProf()
 
+	if o.configPath == "" {
```
Comment (fejta) [inline prow/cmd/plank/main.go]: I would suggest creating a helper method in the config package for this:

```golang

const defaultConfigPath = "/etc/config/config.yaml"

func configPath(value string) string {
  if value != "" {
    return value
  }
  logrus.Warningf("defaulting to %s until whenever, please migrate", defaultConfigPath)
  return defaultConfigPath
}
```

**Comment by cjwagner** [comment]
Please don't include merge commits on your PR branch. Instead, rebase your PR branch onto master and force push the updated PR branch.

File: prow/ANNOUNCEMENTS.md [hunk 4]
```diff
@@ -19,7 +19,9 @@ New features added to each component:
     if you enable postsubmit reporting on.
  - *January 15, 2019* `approve` now considers self-approval and github review
    state by default. Configure with `require_self_approval` and
-   `ignore_review_state`. Temporarily revert to old defaults with `use_deprecated_2018_implicit_self_approve_default_migrate_before_july_2019` and `use_deprecated_2018_review_acts_as_approve_default_migrate_before_july_2019`.
+   `ignore_review_state`. Temporarily revert to old defaults with
```
Comment (fejta) [inline prow/ANNOUNCEMENTS.md]: Any chance you can not change existing lines?

File: prow/config/config.go [hunk 5]
```diff
@@ -818,6 +821,16 @@ func (c *Config) validateJobConfig() error {
 	return nil
 }
 
+// ConfigPath returns the value for the component's configPath if provided
+// explicityly or default otherwise.
+func ConfigPath(value string) string {
+	if value != "" {
+		return value
+	}
+	logrus.Warningf("defaulting to %s until whenever, please migrate", defaultConfigPath)
```
Comment (fejta) [inline prow/config/config.go]: Please pick a date, how about some time in May/June/July

File: prow/config/config.go [hunk 6]
```diff
@@ -53,6 +53,9 @@ const (
 	DefaultJobTimeout = 24 * time.Hour
 )
 
+// TODO: require setting this explicitly
+const defaultConfigPath = "/etc/config/config.yaml"
```
Comment (fejta) [inline prow/config/config.go]: Looks like this is only used by ConfigPath? If so you can scope the const inside the function.

File: prow/config/config.go [hunk 7]
```diff
@@ -818,6 +818,20 @@ func (c *Config) validateJobConfig() error {
...
+// explicityly or default otherwise.
+func ConfigPath(value string) string {
+
+	// TODO: require setting this explicitly
+	defaultConfigPath := "/etc/config/config.yaml"
+
+	if value != "" {
+		return value
+	}
+	logrus.Warningf("defaulting to %s until July 15, please migrate", defaultConfigPath)
```
Comment (fejta) [inline prow/config/config.go]: What year?

Comment (hemani19) [reply prow/config/config.go]: Updated

**Comment by fejta** [comment]
/hold cancel
thanks!

**Comment by stevekuznetsov** [comment]
This ended up being breaking change to the deployments -- we should revert it and try again, NOT changing the defaults until later and only warning on no input.

**Comment by fejta** [comment]
Will you please provide details on what broke? This PR should not have changed any defaults. See thttps://github.com/kubernetes/test-infra/blob/5796ba99ff2c643b8a332e018652f61b075c1067/prow/config/config.go#L821-L833

**Comment by hemanik** [comment]
@stevekuznetsov, @fejta: What should be done with respect to this PR? 

Should I create a new PR for the issue to just add the flgas without changing any defaults or fix validate methods? 

Any suggestions?

**Pull request merged**